README.md
# Pisti

## Table of Contents

- [Pisti](#pisti)
  - [Table of Contents](#table-of-contents)
  - [About](#about)
  - [Requirements](#requirements)
  - [Running on Windows](#running-on-windows)
  - [Built Using](#built-using)
  - [Folder structures](#folder-structures)
  - [Documents](#documents)

## About

Client for Pisti - a Turkey card game.

## Requirements

- Cocos Studio 3.10

- Texture packer

- Node.js

- IDE: Intellij IDEA 2020+

- IMPORTANT: Please read the [Coding Guidelines](./docs/CodingGuidelines.md) before you start coding

## Running on Windows

To run simulator on Windows, run `./run.bat` script from comandline

## Built Using

- Cocos2djs v3.8

## Folder structures

    .
    ├── build                           # Build config folder
    |   ├── ...
    |   ├── vn-debug                    # Configs for debug build for market VN
    |   │   ├── icon                    # Your app icon
    |   │   ├── build.keystore          # Used to build Android APK/ABB
    |   │   ├── keystore_info.json      # Used to build Android APK/ABB
    |   │   ├── cdn.json                # CDN configs
    |   │   ├── config.json             # App config (name, package name, facebook/google things, etc.)
    |   │   ├── google-services.json    # Firebase config file
    |   │   └── version.json            # Version config (android/ios versions, cdn version)
    |   ├── vn-release                  # Configs for release build for market VN (has different CDN to debug build)
    |   ├── ...
    ├── docs                            # Documentation files
    ├── frameworks                      # Frameworks source code: C++, Cocos-html5 API
    ├── res                             # Game assets (sprites, sprite sheets, fonts, configs, etc)
    ├── res-vn                          # Localization assets for market VN (it will be copied into res before build apk - config in cdn.json)
    ├── ...                             # Other market assets
    ├── simulator                       # Win32 exe, dll, JSB files
    ├── src                             # Source files
    ├── tools                           # Tools and utilities
    ├── .cocos-project.json             # Cocos project file
    ├── .gitattributes                  # Used to handle EOF charator in different platform (Windows, Linux, Mac)
    ├── .gitignore                      # Git ignore
    ├── .markdownlint.json              # Config file for markdown lint
    ├── config.json                     # Config for window simulator
    ├── main.js                         # Game entry point
    ├── project.json                    # Cocos project file
    ├── run.bat                         # Run exe, tools script
    ├── CHANGELOG.md                    # Project change log / release note
    └── README.md                       # This file

## Documents

- [Coding Guidelines](./docs/CodingGuidelines.md)


CHANGELOG.md
# Changelog

## [v8.0.2] - 2023/04/07

### Version

- Android version code `22`
- Android version name `v8.0.2`,
- Android cdn version `19`
- Client version `10`

### Updated

- Event Collector tuning ticket buy in shop
- update Event Collector localize

## [v8.0.1] - 2023/04/04

### Version

- Android version code `22`
- Android version name `v8.0.1`,
- Android cdn version `18`
- Client version `9`

### Updated

- Event Collector UI

## [v8.0.0] - 2023/03/30

### Version

- Android version code `21`
- Android version name `v8.0.0`,
- Android cdn version `17`
- Client version `9`

### Added

- Event Collector (POSM)
- Share image

### Updated

- Ramadan theme
- Bigbet system


## [v7.1.0] - 2023/02/23

### Version

- Android version code `20`
- Android version name `v7.1.0`,
- Android cdn version `16`
- Client version `8`

### Added

- Event Pay
- Big Bet

### Updated

- Update minigame animations
- Update Main Menu UI


## [v6.3.1] - 2023/01/17

### Version

- Android cdn version `15`
- Android version name `v6.3.1`,

### Fixed

- Remove some default chats
- Move effect K.O to summary popup

## [v6.3.0] - 2023/01/13

### Version

- Android version code `19`
- Android version name `v6.3.0`,
- Android cdn version `14`
- Client version `7`

### Added

- 30 Days Login Reward
- Offer Webpay (1st) popup

### Updated

- Update reward in offer popup (add pack reward and event tokens)
- Open mailbox after buy offer

## [v6.2.0] - 2023/01/06

### Version

- Android version code `18`
- Android version name `v6.2.0`,
- Android cdn version `13`
- Client version `6`

### Added

- [Pisti] Item factor
- [Event] Dual wheel
- [Sound] Localize voice

### Updated

- [Offer] Auto popup
- [Tutorial] Move guide to setting and hide icon for user lv > 50
- [Minigame/FlipCard] Update anim NPC

## [v6.0.0] - 2022/12/09

### Version

- Android version code `17`
- Android version name `v6.0.0`,
- Android cdn version `12`
- Client version `5`

### Added

- [Payment] Add Papara channel
- [Pisti] Suggest player to join game
- [System] Band user cheat/refund
- [Pisti] Add idle detection & idle auto play

### Updated

- [Offer] Update art
- [Tutorial] Change icon to Guide Book & add flow to show tip when player lose first time
- [Friend] Update art invite friend

## [v5.4.0] - 2022/11/29

### Version

- Android version code `16`
- Android version name `v5.4.0`,
- Android cdn version `11`
- Client version `4`

### Added

- [Payment] Webpay channel
- [Pisti] Auto chat/emoji when player join
- [IOS] Add CDN

### Updated

- Update app to portal
- [Ranking] Art ranking icon

### Fixed

- Crash when tap edit box in Gift Code Popup

## [v5.3.1] - 2022/11/25

### Version

- Android cdn version `10`
- Android version name `v5.3.1`

### Fixed

- [UI] Revert edit box use ck.TextField to ck.EditBox to fix a bug crash when click to edit box

## [v5.3.0] - 2022/11/20

### Version

- Android version code `15`
- Android version name `v5.3.0`,
- Android cdn version `9`
- Client version `3`

### Fixed
- [Rating] Not send info to server when player close popup rating 5 star
- [Item] Not receive double XP when player has X2_XP item
- [Pisti] Stuck when server response swapper used when the game has ended (queue renderer callback not being called)
- [Login] Cache login type (for tracker log)

### Updated
- [Item] UI icon booster in Pisti Scene (countdown real time, tap to view detail info or buy)
- [Login] Remove popup relogin (auto re-login)
- [Tutorial] UI Guide/Tutorial Popup
- [Tutorial] Add text guide in tutorial game
- [Tutorial] After finish tutorial, let user join another game (instead of back to Home Scene)
- [Tutorial] Add Play Now tutorial when user idle at Home Scene
- [BGM] Change Poker BGM, use free music

### Added
- Card collection
- Recommend update
- [Minigames] Add X2 item icon and quick buy layer in Minigames

## [v5.2.3] - 2022/10/31

### Version
- Android cdn version `8`

### Fixed
- [Tutorial] Fix a bug for hot news (type banner and when click).
- [Tutorial] Fix a bug for webview.

## [v5.2.2] - 2022/10/28
### Version
- Android cdn version `7`

### Fixed
- [Tutorial] Fix a bug where player not place random card when timeout.
- [Tutorial] Fix a bug where deck config being changed during tutorial game.

## [v5.2.1] - 2022/10/25

### Version
- Android cdn version `6`

### Fixed
- Fix wrong Ace Heart Image

### Added
- [Poker] Register Exit after Dealing and before Drawing

## [v5.2.0] - 2022/10/21

### Version

- Android version code `14`
- Android version name `v5.0.2`,
- Android cdn version `4`

### Added
- [Pisti] Item tutorial (guide user how to use item)

### Updated
- Update App icon
- [Home] Update UI/UX (Splash, Home, Popups, etc.)
- [Pisti] Update animation/flow waiting state
- [Mini-game/Poker] Update tutorial
- [Item] Change item x2XP type to Booster
- [System] Only update broadcast when in wheel UI
- [Mini-games] Update amount of XP earn from minigame
- [Event/AroundTheWorld] Update fly cost

## [v5.1.0] - 2022/10/05

- [Updated] Update UI
- [Optimized] Using pool for CardNode(s)
- [Optimized] Using cache request (1 request object for all requests)

## [v5.0.0] - 2022/09/09

- Features: Pisti, Quest, Ranking, Free Gifts, Gift Code, Friend, Payment, Shop
- Minigames.
- Events.

src/libs/md5.js
function md5cycle(x, k) {
    var a = x[0], b = x[1], c = x[2], d = x[3];

    a = ff(a, b, c, d, k[0], 7, -680876936);
    d = ff(d, a, b, c, k[1], 12, -389564586);
    c = ff(c, d, a, b, k[2], 17, 606105819);
    b = ff(b, c, d, a, k[3], 22, -1044525330);
    a = ff(a, b, c, d, k[4], 7, -176418897);
    d = ff(d, a, b, c, k[5], 12, 1200080426);
    c = ff(c, d, a, b, k[6], 17, -1473231341);
    b = ff(b, c, d, a, k[7], 22, -45705983);
    a = ff(a, b, c, d, k[8], 7, 1770035416);
    d = ff(d, a, b, c, k[9], 12, -1958414417);
    c = ff(c, d, a, b, k[10], 17, -42063);
    b = ff(b, c, d, a, k[11], 22, -1990404162);
    a = ff(a, b, c, d, k[12], 7, 1804603682);
    d = ff(d, a, b, c, k[13], 12, -40341101);
    c = ff(c, d, a, b, k[14], 17, -1502002290);
    b = ff(b, c, d, a, k[15], 22, 1236535329);

    a = gg(a, b, c, d, k[1], 5, -165796510);
    d = gg(d, a, b, c, k[6], 9, -1069501632);
    c = gg(c, d, a, b, k[11], 14, 643717713);
    b = gg(b, c, d, a, k[0], 20, -373897302);
    a = gg(a, b, c, d, k[5], 5, -701558691);
    d = gg(d, a, b, c, k[10], 9, 38016083);
    c = gg(c, d, a, b, k[15], 14, -660478335);
    b = gg(b, c, d, a, k[4], 20, -405537848);
    a = gg(a, b, c, d, k[9], 5, 568446438);
    d = gg(d, a, b, c, k[14], 9, -1019803690);
    c = gg(c, d, a, b, k[3], 14, -187363961);
    b = gg(b, c, d, a, k[8], 20, 1163531501);
    a = gg(a, b, c, d, k[13], 5, -1444681467);
    d = gg(d, a, b, c, k[2], 9, -51403784);
    c = gg(c, d, a, b, k[7], 14, 1735328473);
    b = gg(b, c, d, a, k[12], 20, -1926607734);

    a = hh(a, b, c, d, k[5], 4, -378558);
    d = hh(d, a, b, c, k[8], 11, -2022574463);
    c = hh(c, d, a, b, k[11], 16, 1839030562);
    b = hh(b, c, d, a, k[14], 23, -35309556);
    a = hh(a, b, c, d, k[1], 4, -1530992060);
    d = hh(d, a, b, c, k[4], 11, 1272893353);
    c = hh(c, d, a, b, k[7], 16, -155497632);
    b = hh(b, c, d, a, k[10], 23, -1094730640);
    a = hh(a, b, c, d, k[13], 4, 681279174);
    d = hh(d, a, b, c, k[0], 11, -358537222);
    c = hh(c, d, a, b, k[3], 16, -722521979);
    b = hh(b, c, d, a, k[6], 23, 76029189);
    a = hh(a, b, c, d, k[9], 4, -640364487);
    d = hh(d, a, b, c, k[12], 11, -421815835);
    c = hh(c, d, a, b, k[15], 16, 530742520);
    b = hh(b, c, d, a, k[2], 23, -995338651);

    a = ii(a, b, c, d, k[0], 6, -198630844);
    d = ii(d, a, b, c, k[7], 10, 1126891415);
    c = ii(c, d, a, b, k[14], 15, -1416354905);
    b = ii(b, c, d, a, k[5], 21, -57434055);
    a = ii(a, b, c, d, k[12], 6, 1700485571);
    d = ii(d, a, b, c, k[3], 10, -1894986606);
    c = ii(c, d, a, b, k[10], 15, -1051523);
    b = ii(b, c, d, a, k[1], 21, -2054922799);
    a = ii(a, b, c, d, k[8], 6, 1873313359);
    d = ii(d, a, b, c, k[15], 10, -30611744);
    c = ii(c, d, a, b, k[6], 15, -1560198380);
    b = ii(b, c, d, a, k[13], 21, 1309151649);
    a = ii(a, b, c, d, k[4], 6, -145523070);
    d = ii(d, a, b, c, k[11], 10, -1120210379);
    c = ii(c, d, a, b, k[2], 15, 718787259);
    b = ii(b, c, d, a, k[9], 21, -343485551);

    x[0] = add32(a, x[0]);
    x[1] = add32(b, x[1]);
    x[2] = add32(c, x[2]);
    x[3] = add32(d, x[3]);

}

function cmn(q, a, b, x, s, t) {
    a = add32(add32(a, q), add32(x, t));
    return add32((a << s) | (a >>> (32 - s)), b);
}

function ff(a, b, c, d, x, s, t) {
    return cmn((b & c) | ((~b) & d), a, b, x, s, t);
}

function gg(a, b, c, d, x, s, t) {
    return cmn((b & d) | (c & (~d)), a, b, x, s, t);
}

function hh(a, b, c, d, x, s, t) {
    return cmn(b ^ c ^ d, a, b, x, s, t);
}

function ii(a, b, c, d, x, s, t) {
    return cmn(c ^ (b | (~d)), a, b, x, s, t);
}

function md51(s) {
    txt = '';
    var n = s.length,
        state = [1732584193, -271733879, -1732584194, 271733878], i;
    for (i = 64; i <= s.length; i += 64) {
        md5cycle(state, md5blk(s.substring(i - 64, i)));
    }
    s = s.substring(i - 64);
    var tail = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
    for (i = 0; i < s.length; i++)
        tail[i >> 2] |= s.charCodeAt(i) << ((i % 4) << 3);
    tail[i >> 2] |= 0x80 << ((i % 4) << 3);
    if (i > 55) {
        md5cycle(state, tail);
        for (i = 0; i < 16; i++) tail[i] = 0;
    }
    tail[14] = n * 8;
    md5cycle(state, tail);
    return state;
}

/* there needs to be support for Unicode here,
 * unless we pretend that we can redefine the MD-5
 * algorithm for multi-byte characters (perhaps
 * by adding every four 16-bit characters and
 * shortening the sum to 32 bits). Otherwise
 * I suggest performing MD-5 as if every character
 * was two bytes--e.g., 0040 0025 = @%--but then
 * how will an ordinary MD-5 sum be matched?
 * There is no way to standardize text to something
 * like UTF-8 before transformation; speed cost is
 * utterly prohibitive. The JavaScript standard
 * itself needs to look at this: it should start
 * providing access to strings as preformed UTF-8
 * 8-bit unsigned value arrays.
 */
function md5blk(s) { /* I figured global was faster.   */
    var md5blks = [], i;
    /* Andy King said do it this way. */
    for (i = 0; i < 64; i += 4) {
        md5blks[i >> 2] = s.charCodeAt(i)
            + (s.charCodeAt(i + 1) << 8)
            + (s.charCodeAt(i + 2) << 16)
            + (s.charCodeAt(i + 3) << 24);
    }
    return md5blks;
}

var hex_chr = '0123456789abcdef'.split('');

function rhex(n) {
    var s = '', j = 0;
    for (; j < 4; j++)
        s += hex_chr[(n >> (j * 8 + 4)) & 0x0F]
            + hex_chr[(n >> (j * 8)) & 0x0F];
    return s;
}

function hex(x) {
    for (var i = 0; i < x.length; i++)
        x[i] = rhex(x[i]);
    return x.join('');
}

function md5(s) {
    return hex(md51(s));
}

/* this function is much faster,
so if possible we use it. Some IEs
are the only ones I know of that
need the idiotic second function,
generated by an if clause.  */

function add32(a, b) {
    return (a + b) & 0xFFFFFFFF;
}

if (md5('hello') != '5d41402abc4b2a76b9719d911017c592') {
    function add32(x, y) {
        var lsw = (x & 0xFFFF) + (y & 0xFFFF),
            msw = (x >> 16) + (y >> 16) + (lsw >> 16);
        return (msw << 16) | (lsw & 0xFFFF);
    }
}
